File: core-service/src/main/java/com/stride/tracking/coreservice/service/impl/MapboxService.java
package com.stride.tracking.coreservice.service.impl;

import com.stride.tracking.bridge.dto.supabase.response.FileLinkResponse;
import com.stride.tracking.commons.utils.FeignClientHandler;
import com.stride.tracking.core.dto.mapbox.response.MapboxDirectionResponse;
import com.stride.tracking.core.dto.mapbox.response.MapboxWayPoint;
import com.stride.tracking.core.dto.sport.SportMapType;
import com.stride.tracking.coreservice.client.BridgeFeignClient;
import com.stride.tracking.coreservice.client.MapboxFeignClient;
import com.stride.tracking.coreservice.constant.Message;
import com.stride.tracking.coreservice.utils.JsonHelper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class MapboxService {
    private final MapboxFeignClient mapboxClient;
    private final BridgeFeignClient bridgeClient;

    @Value("${mapbox.access-token}")
    private String accessToken;

    @Value("${mapbox.static-image.style}")
    private String staticImageDefaultStyle;

    @Value("${mapbox.static-image.stroke-width}")
    private int staticImageDefaultStrokeWidth;

    @Value("${mapbox.static-image.stroke-color}")
    private String staticImageDefaultStrokeColor;

    @Value("${mapbox.static-image.stroke-fill}")
    private String staticImageDefaultStrokeFill;

    @Value("${mapbox.static-image.width}")
    private int staticImageDefaultWidth;

    @Value("${mapbox.static-image.height}")
    private int staticImageDefaultHeight;

    @Value("${mapbox.static-image.padding}")
    private int staticImageDefaultPadding;

    @Value("${mapbox.static-image.content-type}")
    private String staticImageContentType;

    @Value("${mapbox.directions.alternatives}")
    private String directionsAlternatives;

    @Value("${mapbox.directions.geometries}")
    private String directionsGeometries;

    @Value("${mapbox.directions.overview}")
    private String directionsOverview;

    @Value("${mapbox.directions.steps}")
    private String directionsSteps;

    @Value("${mapbox.directions.continue_straight}")
    private String directionsContinueStraight;

    @Value("${mapbox.matching.geometries}")
    private String matchingGeometries;

    @Value("${mapbox.matching.overview}")
    private String matchingOverview;

    public MapboxDirectionResponse getBatchRoute(List<List<Double>> coordinates, SportMapType mapType) {
        int maxCoordsPerRequest = 25;
        List<List<List<Double>>> chunks = new ArrayList<>();

        for (int i = 0; i < coordinates.size(); i += maxCoordsPerRequest) {
            int end = Math.min(i + maxCoordsPerRequest, coordinates.size());
            chunks.add(coordinates.subList(i, end));
        }

        List<List<Double>> allCoordinates = new ArrayList<>();
        List<MapboxWayPoint> allWaypoints = new ArrayList<>();

        for (List<List<Double>> chunk : chunks) {
            MapboxDirectionResponse response = getRoute(chunk, mapType);
            allCoordinates.addAll(response.getCoordinates());
            allWaypoints.addAll(response.getWayPoints());
        }

        Map<String, MapboxWayPoint> waypointMap = new HashMap<>();
        for (MapboxWayPoint wp : allWaypoints) {
            String key = wp.getName().toLowerCase();

            waypointMap.compute(key, (k, v) -> {
                if (v == null) {
                    return MapboxWayPoint.builder()
                            .latitude(wp.getLatitude())
                            .longitude(wp.getLongitude())
                            .name(wp.getName())
                            .freq(wp.getFreq())
                            .build();
                }
                v.setFreq(v.getFreq() + wp.getFreq());
                return v;
            });
        }

        List<MapboxWayPoint> mergedWaypoints = new ArrayList<>(waypointMap.values());
        return new MapboxDirectionResponse(allCoordinates, mergedWaypoints);
    }

    public MapboxDirectionResponse getRoute(List<List<Double>> coordinates, SportMapType mapType) {
        String encodedCoords = encodeCoords(coordinates);

        Map<String, Object> response = FeignClientHandler.handleExternalCall(
                () -> mapboxClient.getDirections(
                        mapType.name().toLowerCase(), encodedCoords, accessToken,
                        directionsAlternatives, directionsGeometries, directionsOverview,
                        directionsSteps, directionsContinueStraight
                ),
                HttpStatus.INTERNAL_SERVER_ERROR,
                Message.CAN_NOT_GET_DIRECTIONS_FROM_MAPBOX
        );

        List<List<Double>> coords = (List<List<Double>>) JsonHelper.getNestedValue(response, "routes", "0", "geometry", "coordinates");
        List<Map<String, Object>> rawWps = (List<Map<String, Object>>) response.getOrDefault("waypoints", new ArrayList<>());

        Map<String, MapboxWayPoint> counter = new HashMap<>();
        for (Map<String, Object> wp : rawWps) {
            String name = (String) wp.get("name");
            if (name == null || name.isEmpty()) {
                continue;
            }

            if (counter.containsKey(name)) {
                MapboxWayPoint mapboxWayPoint = counter.get(name);
                mapboxWayPoint.setFreq(mapboxWayPoint.getFreq() + 1);
            } else {
                counter.put(
                        name,
                        MapboxWayPoint.builder()
                                .name(name)
                                .latitude((Double) JsonHelper.getNestedValue(wp, "location", "1"))
                                .longitude((Double) JsonHelper.getNestedValue(wp, "location", "0"))
                                .freq(1)
                                .build()
                );
            }
        }

        List<MapboxWayPoint> waypoints = new ArrayList<>(counter.values());

        waypoints.sort(Comparator.comparingInt(MapboxWayPoint::getFreq).reversed());

        return new MapboxDirectionResponse(coords, waypoints);
    }

    private String encodeCoords(List<List<Double>> coordinates) {
        return coordinates.stream()
                .map(coordinate -> String.format("%s,%s", coordinate.get(0), coordinate.get(1)))
                .collect(Collectors.joining(";"));
    }

    public String generateAndUpload(String path, String fileName) {
        byte[] imageData = generateImage(path);

        return uploadFile(imageData, fileName);
    }

    private byte[] generateImage(String path) {
        return generateImage(
                path,
                staticImageDefaultStyle,
                staticImageDefaultStrokeWidth,
                staticImageDefaultStrokeColor,
                staticImageDefaultStrokeFill,
                staticImageDefaultWidth,
                staticImageDefaultHeight,
                staticImageDefaultPadding
        );
    }

    private byte[] generateImage(
            String path,
            String style,
            int strokeWidth,
            String strokeColor,
            String strokeFill,
            int width,
            int height,
            int padding
    ) {
        return FeignClientHandler.handleExternalCall(
                () -> mapboxClient.getStaticMapImage(
                        style, strokeWidth, strokeColor, strokeFill, path, width, height, padding, accessToken
                ),
                HttpStatus.INTERNAL_SERVER_ERROR,
                Message.GENERATE_IMAGE_FAILED
        );
    }

    private String uploadFile(byte[] data, String name) {
        FileLinkResponse response = FeignClientHandler.handleInternalCall(
                () -> bridgeClient.uploadRawFile(data, name, staticImageContentType),
                HttpStatus.INTERNAL_SERVER_ERROR,
                Message.UPLOAD_IMAGE_FAILED
        );
        return response.getFile();
    }

    public List<List<Double>> getBatchMapMatchingPoints(SportMapType mapType, List<List<Double>> coordinates) {
        int maxCoordsPerRequest = 100;
        List<List<List<Double>>> chunks = new ArrayList<>();

        for (int i = 0; i < coordinates.size(); i += maxCoordsPerRequest) {
            int end = Math.min(i + maxCoordsPerRequest, coordinates.size());
            chunks.add(coordinates.subList(i, end));
        }

        List<List<Double>> allCoordinates = new ArrayList<>();

        for (List<List<Double>> chunk : chunks) {
            List<List<Double>> response = getMapMatchingPoints(mapType, chunk);
            allCoordinates.addAll(response);
        }

        return allCoordinates;
    }

    public List<List<Double>> getMapMatchingPoints(SportMapType mapType, List<List<Double>> coordinates) {
        String encodedCoords = encodeCoords(coordinates);

        Map<String, Object> response = FeignClientHandler.handleExternalCall(
                () -> mapboxClient.getMatching(
                        mapType.name().toLowerCase(),
                        encodedCoords,
                        accessToken,
                        matchingGeometries,
                        matchingOverview
                ),
                HttpStatus.INTERNAL_SERVER_ERROR,
                Message.CAN_NOT_GET_DIRECTIONS_FROM_MAPBOX
        );

        List<Object> tracePointsRaw = (List<Object>) response.getOrDefault("tracepoints", new ArrayList<>());
        List<List<Double>> result = new ArrayList<>();

        for (Object tpObj : tracePointsRaw) {
            if (tpObj instanceof Map) {
                Map<String, Object> tracePoint = (Map<String, Object>) tpObj;
                List<Double> location = (List<Double>) tracePoint.get("location");
                result.add(location);
            }
        }

        return result;
    }
}

File: route-service/services/mapbox_service.py
from urllib.parse import quote
from collections import defaultdict
from typing import List

import requests
from fastapi import status

from clients.bridge_client import BridgeClient
from clients.mapbox_client import MapboxClient
from configuration.manager import settings
from constants.geometry_encode_type import GeometryEncodeType
from constants.map_type import SportMapType
from dto.mapbox.response.mapbox_direction_response import MapboxWayPoint, MapboxDirectionResponse
from exceptions.common_exception import StrideException
from utils.data_helper import DataHelper
from utils.geometry_helper import GeometryHelper


class MapboxService:
    def __init__(self, mapbox_client: MapboxClient, bridge_client: BridgeClient):
        self.mapbox_client = mapbox_client
        self.bridge_client = bridge_client
        self.default_style = settings.MAPBOX_STYLE
        self.default_stroke_width = settings.MAPBOX_STROKE_WIDTH
        self.default_stroke_color = settings.MAPBOX_STROKE_COLOR
        self.default_stroke_fill = settings.MAPBOX_STROKE_FILL
        self.default_width = settings.MAPBOX_WIDTH
        self.default_height = settings.MAPBOX_HEIGHT
        self.default_padding = settings.MAPBOX_PADDING
        self.content_type = settings.MAPBOX_CONTENT_TYPE

    def get_batch_route(
            self,
            coordinates: List[List[float]],
            map_type: SportMapType = SportMapType.DRIVING
    ) -> MapboxDirectionResponse:
        max_coords_per_request = 25
        coord_chunks = [coordinates[i:i + max_coords_per_request] for i in
                        range(0, len(coordinates), max_coords_per_request)]

        combined_coordinates, combined_waypoints = [], []
        for chunk in coord_chunks:
            result = self.get_route(chunk, map_type)
            combined_coordinates.extend(result.coordinates)
            combined_waypoints.extend(result.waypoints)

        # Group waypoints by name
        waypoint_dict = defaultdict(lambda: {
            "latitude": 0.0,
            "longitude": 0.0,
            "freq": 0,
            "name": ""
        })
        for wp in combined_waypoints:
            key = wp.name.lower()
            waypoint_dict[key]["latitude"] = wp.latitude
            waypoint_dict[key]["longitude"] = wp.longitude
            waypoint_dict[key]["freq"] += wp.freq
            waypoint_dict[key]["name"] = wp.name

        waypoints = [MapboxWayPoint(**val) for val in waypoint_dict.values()]
        return MapboxDirectionResponse(coordinates=combined_coordinates, waypoints=waypoints)

    def get_route(
            self,
            coordinates: List[List[float]],
            map_type: SportMapType = SportMapType.DRIVING
    ) -> MapboxDirectionResponse:
        encoded_coords = GeometryHelper.encode_geometry(
            coordinates=coordinates,
            encode_type=GeometryEncodeType.URL
        )
        response = self.mapbox_client.get_directions(
            map_type=map_type.lowercase,
            coordinates=encoded_coords
        )

        coords = DataHelper.safe_get_nested(
            response,
            ["routes", 0, "geometry", "coordinates"],
            default=[]
        )
        raw_wps = response.get("waypoints", [])
        counter = {}
        for wp in raw_wps:
            name = wp.get("name", "")
            if not name:
                continue
            if name in counter:
                counter[name]["freq"] += 1
            else:
                counter[name] = {
                    "latitude": wp["location"][1],
                    "longitude": wp["location"][0],
                    "freq": 1,
                }

        waypoints = [MapboxWayPoint(
            name=k,
            latitude=v["latitude"],
            longitude=v["longitude"],
            freq=v["freq"]
        ) for k, v in counter.items()]
        waypoints.sort(key=lambda w: w.freq, reverse=True)
        return MapboxDirectionResponse(coordinates=coords, waypoints=waypoints)

    def generate_and_upload(self, path: str, file_name: str) -> str:
        image_data = self._generate_image(path)
        return self._upload_file(image_data, file_name)

    def _generate_image(self, path: str) -> bytes:
        return self._generate_image_with_params(
            path,
            self.default_style,
            self.default_stroke_width,
            self.default_stroke_color,
            self.default_stroke_fill,
            self.default_width,
            self.default_height,
            self.default_padding,
        )

    def _generate_image_with_params(
            self,
            path: str,
            style: str,
            stroke_width: str,
            stroke_color: str,
            stroke_fill: str,
            width: int,
            height: int,
            padding: int
    ) -> bytes:
        try:
            encoded_path = quote(path.encode('utf-8'), safe="")

            return self.mapbox_client.get_static_map_image(
                map_style=style,
                stroke_width=stroke_width,
                stroke_color=stroke_color,
                stroke_fill=stroke_fill,
                path=encoded_path,
                width=width,
                height=height,
                padding=padding
            )
        except requests.HTTPError as e:
            raise StrideException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                message=f"Failed to generate image: {e}"
            )

    def _upload_file(self, data: bytes, name: str) -> str:
        try:
            response = self.bridge_client.upload_raw_file(
                data=data,
                file_name=name,
                content_type=self.content_type
            )
            return response.file
        except requests.HTTPError as e:
            raise StrideException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                message=f"Failed to upload image: {e}"
            )
